src/mpp/soe/SystemOfEquationsBaseType.F90

```diff
@@ -586,6 +589,28 @@ subroutine SOEBaseStepDT_KSP(soe, dt, converged, ierr)
     ! Compute the 'b' vector
     call soe%ComputeRHS(soe%solver%ksp, soe%solver%rhs, ierr)
 
+    ! Do precise preallocation for the matrix 'A' using MATPREALLOCATOR.
+    ! There is no easy way to accurately determine the needed preallocation a priori, so we use
+    ! use MATPREALLOCATOR to count. This is fine, as hash tables are fast!
+
+    ! Note: We really should do the MatCreate() call as below, but PetscObjectComm() is currently
+    ! unsupported in Fortran. So we are cheating and using PETSC_COMM_WORLD, since currently no
+    ! other communicator is used. We could also use PETSC_COMM_SELF, as currently we also only run
+    ! this on a single rank. But these things may change, so it would be better to get the communicator
+    ! directly from soe%solver%Amat.
+    ! call MatCreate(PetscObjectComm(soe%solver%Amat), preallocator, ierr); CHKERRQ(ierr)
+    call MatCreate(PETSC_COMM_WORLD, preallocator, ierr); CHKERRQ(ierr)
```

Comment:
When MPP lib is coupled with ELM, each rank independently calls MPP. So, let's change `PETSC_COMM_WORLD` to `PETSC_COMM_SELF`. What do you think?

Reply:
Hi @bishtgautam. Apologies for not replying to your comment for months, which I didn't notice until now (I need to mess with my Github notification settings, I guess!). As my comment says, really what I should do is make `PetscObjectComm()` work from Fortran and use the communicator that that returns. The way the code is being run now, though, we could use either `PETSC_COMM_WORLD` or `PETSC_COMM_SELF`. What I wonder about is what `PETSC_COMM_WORLD` should be in the case when MPP lib is coupled with ELM. If `PETSC_COMM_WORLD` is going to be the same as `MPI_COMM_WORLD`, then we need to use `PETSC_COMM_SELF` here. But in your usage model from ELM, it seems like the "correct" thing to do might be to set `PETSC_COMM_WORLD` to `MPI_COMM_SELF`. This would leave open the possibility that, sometime in the future, perhaps we might want to run MPP on a communicator that has size greater than one (but still some subset of `MPI_COMM_WORLD`). Keeping this bit of flexibility seems a little better than hard-coding things inside MPP lib's PETSc calls to assume that we are always using only a single rank. But it's your call as to how you want to do things from ELM.